trlx/trainer/accelerate_ppo_trainer.py

```diff
@@ -286,6 +286,42 @@ def make_experience(self, num_rollouts: int = 1024, iter_count: int = 0):  # noq
 
             prompt_tensors = batch.input_ids
             device = samples.device
+
+            prompt_sizes = torch.tensor([prompt_tensors.shape[1]] * len(prompt_tensors), device=device)
+            padded_samples = self.accelerator.pad_across_processes(
+                samples, dim=1, pad_index=self.tokenizer.eos_token_id, pad_first=False
+            )
+            padded_prompts = self.accelerator.pad_across_processes(
+                prompt_tensors, dim=1, pad_index=self.tokenizer.eos_token_id, pad_first=False
+            )
+            gathered_samples = self.accelerator.gather(padded_samples)
+            gathered_prompts = self.accelerator.gather(padded_prompts)
+            gathered_prompt_sizes = self.accelerator.gather(prompt_sizes)
+
+            if torch.distributed.get_rank() == 0:
+                all_str_samples, all_str_prompts, all_str_outputs = self.decode(
+                    gathered_prompts, gathered_samples, gathered_prompt_sizes
+                )
+
+                exp_score_time = time()
+                all_scores = torch.tensor(
+                    self.reward_fn(
+                        samples=all_str_samples,
+                        prompts=all_str_prompts,
+                        outputs=all_str_outputs,
+                    ),
+                    dtype=torch.float,
+                    device=device,
+                )
+                stats["time/exp_score"] = time() - exp_score_time
+
+                all_scores = list(all_scores.reshape(torch.distributed.get_world_size(), -1).unbind())
+            else:
+                all_scores = None
+
+            scores = torch.empty(len(samples), device=device)
+            torch.distributed.scatter(scores, all_scores)
+
             str_samples, str_prompts, str_outputs = self.decode(prompt_tensors, samples)
```

Comment:
To avoid repetetion I could scatter samples as well, but I don't think it's worth it since the second `decode` (which would be second only on the main rank) is needed just to strip `stop_sequences`, so it's basically just an allocation. Runtime in `make_experience` is still dominated by `generate` and `reward_fn`

Reply:
Yeah, that's fine! (I actually meant to approve this last night but forgot to 😅)